### PR TITLE
Support sql annotation analysis

### LINKED ISSUE #12279: Support sql annotation analysis.
Support sql annotation analysis.
``` -- begin comments\n Select * FROM t_order;```
After paser this sql, we can get comments segment `-- begin comments\n` in `SQLStatement`

### PR DESCRIPTION
**Pull request by tuichenchuxin**
Fixes #12279.

Changes proposed in this pull request:
- support sql annotation analysis.
- add test case.

### CODE REVIEW

File: shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/jaxb/cases/domain/segment/impl/comments/ExpectedComments.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+import javax.xml.bind.annotation.XmlAttribute;
+
+/**
+ * Expected comments.
+ */
+@Getter
+@Setter
+public class ExpectedComments extends AbstractExpectedSQLSegment {
+    @XmlAttribute
```
Comment (taojintianxia) [inline shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/jaxb/cases/domain/segment/impl/comments/ExpectedComments.java:32]: leave an empty line between the class clarification and XmlAttribute annotation, like :

```java
public class ExpectedComments extends AbstractExpectedSQLSegment {
    
    @XmlAttribute
```

Comment (tuichenchuxin) [reply shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/jaxb/cases/domain/segment/impl/comments/ExpectedComments.java:32]: ok, thanks.

> leave an empty line between the class clarification and XmlAttribute annotation, like :
> 
> ```java
> public class ExpectedComments extends AbstractExpectedSQLSegment {
>     
>     @XmlAttribute
> ```

ok, thanks for reminding.

File: shardingsphere-infra/shardingsphere-infra-parser/src/main/java/org/apache/shardingsphere/infra/parser/sql/SQLStatementParserExecutor.java [hunk 2]
```diff
@@ -44,6 +48,16 @@ public SQLStatementParserExecutor(final String databaseType) {
...
-        return visitorEngine.visit(parserEngine.parse(sql, false));
+        ParserContext parserContext = parserEngine.parse(sql, false);
+        SQLStatement result = visitorEngine.visit(parserContext.getParseTree());
+        handleComments(result, parserContext);
+        return result;
+    }
+    
+    private void handleComments(final SQLStatement sqlStatement, final ParserContext parserContext) {
+        if (sqlStatement instanceof AbstractSQLStatement) {
+            ((AbstractSQLStatement) sqlStatement).setCommentsSegments(parserContext.getHiddenTokens().stream()
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-parser/src/main/java/org/apache/shardingsphere/infra/parser/sql/SQLStatementParserExecutor.java:59]: @tuichenchuxin Can we use addAll method here?

File: shardingsphere-infra/shardingsphere-infra-parser/src/main/java/org/apache/shardingsphere/infra/parser/sql/SQLStatementParserExecutor.java [hunk 3]
```diff
@@ -44,6 +48,16 @@ public SQLStatementParserExecutor(final String databaseType) {
      * @return SQL statement
      */
     public SQLStatement parse(final String sql) {
-        return visitorEngine.visit(parserEngine.parse(sql, false));
+        ParserContext parserContext = parserEngine.parse(sql, false);
+        SQLStatement result = visitorEngine.visit(parserContext.getParseTree());
+        handleComments(result, parserContext);
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-parser/src/main/java/org/apache/shardingsphere/infra/parser/sql/SQLStatementParserExecutor.java:53]: @tuichenchuxin Do you think `appendSQLComments` is better for this method?

File: shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/main/java/org/apache/shardingsphere/sql/parser/core/ParseASTNode.java [hunk 4]
```diff
@@ -17,18 +17,34 @@
...
+    private final Collection<Token> hiddenTokens;
+    
+    public ParseASTNode(final ParseTree parseTree) {
+        this.parseTree = parseTree;
+        this.hiddenTokens = new LinkedList<>();
+    }
+    
+    public ParseASTNode(final ParseTree parseTree, final CommonTokenStream tokenStream) {
+        this.parseTree = parseTree;
+        this.hiddenTokens = tokenStream.getTokens().stream().filter(each -> each.getChannel() == Token.HIDDEN_CHANNEL).collect(Collectors.toList());
```
Comment (strongduanmu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/main/java/org/apache/shardingsphere/sql/parser/core/ParseASTNode.java:45]: @tuichenchuxin Please place the constant to the left of the expression.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/main/java/org/apache/shardingsphere/sql/parser/core/ParserContext.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import org.antlr.v4.runtime.Token;
+import org.antlr.v4.runtime.tree.ParseTree;
+
+import java.util.Collection;
+
+@RequiredArgsConstructor
+@Getter
+public class ParserContext {
```
Comment (strongduanmu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/main/java/org/apache/shardingsphere/sql/parser/core/ParserContext.java:29]: @tuichenchuxin Add final here.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/generic/CommentsSegment.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
+/**
+ * Comments segment.
+ */
+@RequiredArgsConstructor
+@Getter
+public class CommentsSegment {
```
Comment (strongduanmu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/generic/CommentsSegment.java:28]: @tuichenchuxin Add final for this class and extend SQLStatement.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/generic/CommentSegment.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
+/**
+ * Comment segment.
+ */
+@RequiredArgsConstructor
+@Getter
+public final class CommentSegment {
```
Comment (strongduanmu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/generic/CommentSegment.java:28]: @tuichenchuxin Please extends SQLSegment interface.

File: shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/asserts/statement/SQLStatementAssert.java [hunk 8]
```diff
@@ -49,6 +56,9 @@
      */
     public static void assertIs(final SQLCaseAssertContext assertContext, final SQLStatement actual, final SQLParserTestCase expected) {
         ParameterMarkerAssert.assertCount(assertContext, actual.getParameterCount(), expected.getParameters().size());
+        if (!expected.getComments().isEmpty()) {
```
Comment (strongduanmu) [inline shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/asserts/statement/SQLStatementAssert.java:59]: @tuichenchuxin Can we add a `CommentAssert` to handle this logic? And move `if (!expected.getComments().isEmpty())` into the `assertComment` method.

File: shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/jaxb/cases/domain/statement/SQLParserTestCase.java [hunk 9]
```diff
@@ -41,6 +44,9 @@
     @XmlAttribute
     private String parameters;
     
+    @XmlElement(name = "comments")
+    private List<ExpectedComment> comments = new LinkedList<>();
```
Comment (strongduanmu) [inline shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/jaxb/cases/domain/statement/SQLParserTestCase.java:48]: @tuichenchuxin `@XmlElement(name = "comment")` may be better.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

**Review by strongduanmu — APPROVED** [review]

File: shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/asserts/statement/comment/CommentAssert.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Comment assert.
+ */
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class CommentAssert {
```
Comment (tristaZero) [inline shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/asserts/statement/comment/CommentAssert.java:39]: final class?

**Review by tristaZero — CHANGES_REQUESTED** [review]

**Comment by tristaZero** [comment]
@tuichenchuxin @taojintianxia Do we have any performance test result for this PR?

**Comment by tuichenchuxin** [comment]
> @tuichenchuxin @taojintianxia Do we have any performance test result for this PR?

@tristaZero @taojintianxia had tested performance. Paser+visitor slow down 1%-3%. Only parser (SQLParserEngine) slow down 7%-8%. This performance test is similar to mine.
Detail:https://u01f1kqxrl.feishu.cn/sheets/shtcnp869AyY2pZDemMWwQOB9sg

File: shardingsphere-distsql-parser/shardingsphere-distsql-parser-engine/src/main/java/org/apache/shardingsphere/distsql/parser/core/advanced/AdvancedDistSQLParser.java [hunk 11]
```diff
@@ -23,12 +23,14 @@
...
 
+import java.util.Properties;
+
 /**
  * SQL parser for advanced dist SQL.
  */
 public final class AdvancedDistSQLParser extends AdvancedDistSQLStatementParser implements SQLParser {
     
-    public AdvancedDistSQLParser(final TokenStream input) {
+    public AdvancedDistSQLParser(final TokenStream input, final Properties props) {
```
Comment (strongduanmu) [inline shardingsphere-distsql-parser/shardingsphere-distsql-parser-engine/src/main/java/org/apache/shardingsphere/distsql/parser/core/advanced/AdvancedDistSQLParser.java:33]: @tuichenchuxin Can we remove this useless props?

File: shardingsphere-sql-parser/shardingsphere-sql-parser-spi/src/main/java/org/apache/shardingsphere/sql/parser/api/parser/SQLParser.java [hunk 12]
```diff
@@ -24,6 +24,8 @@
  */
 public interface SQLParser {
     
+    String ENABLE_SQL_COMMENT_PARSE = "enable-sql-comment-parse";
```
Comment (strongduanmu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-spi/src/main/java/org/apache/shardingsphere/sql/parser/api/parser/SQLParser.java:27]: @tuichenchuxin Can we replace this constant with ConfigurationPropertyKey?

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/execute/MySQLComStmtExecuteExecutor.java [hunk 13]
```diff
@@ -68,7 +68,8 @@
     public MySQLComStmtExecuteExecutor(final MySQLComStmtExecutePacket packet, final BackendConnection backendConnection) {
         String schemaName = backendConnection.getSchemaName();
         ShardingSphereSQLParserEngine sqlStatementParserEngine = new ShardingSphereSQLParserEngine(DatabaseTypeRegistry.getTrunkDatabaseTypeName(
-                ProxyContext.getInstance().getContextManager().getMetaDataContexts().getMetaData(schemaName).getResource().getDatabaseType()));
+                ProxyContext.getInstance().getContextManager().getMetaDataContexts().getMetaData(schemaName).getResource().getDatabaseType()),
```
Comment (strongduanmu) [inline shardingsphere-proxy/shardingsphere-proxy-frontend/shardingsphere-proxy-frontend-mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/command/query/binary/execute/MySQLComStmtExecuteExecutor.java:71]: @tuichenchuxin Can we extract param MetaDataContexts?

File: shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/main/java/org/apache/shardingsphere/sql/parser/core/SQLParserFactory.java [hunk 14]
```diff
@@ -45,19 +47,35 @@
...
     @SneakyThrows(ReflectiveOperationException.class)
-    private static SQLParser createSQLParser(final TokenStream tokenStream, final Class<? extends SQLParser> parserClass) {
+    private static SQLParser createSQLParser(final TokenStream tokenStream, final Class<? extends SQLParser> parserClass, final boolean sqlCommentParseEnabled) {
         SQLParser result = parserClass.getConstructor(TokenStream.class).newInstance(tokenStream);
+        setEnableSqlCommentParse(result, parserClass, sqlCommentParseEnabled);
         ((Parser) result).setErrorHandler(new BailErrorStrategy());
         return result;
     }
     
+    private static void setEnableSqlCommentParse(final SQLParser result, final Class<? extends SQLParser> parserClass, final boolean sqlCommentParseEnabled)
```
Comment (strongduanmu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/main/java/org/apache/shardingsphere/sql/parser/core/SQLParserFactory.java:65]: @tuichenchuxin Can this method be simplified?

**Review by strongduanmu — CHANGES_REQUESTED** [review]

**Review by strongduanmu — APPROVED** [review]

**Review by tristaZero — APPROVED** [review]

**Pull request merged**